File: lib/ui/screens/album_detail_screen.dart
import 'package:audio_service/audio_service.dart';
import 'package:flutter/material.dart';
import 'package:flutter_cache_manager/flutter_cache_manager.dart';
import 'package:gap/gap.dart';
import 'package:groovybox/data/db.dart';
import 'package:groovybox/data/playlist_repository.dart';
import 'package:groovybox/providers/audio_provider.dart';
import 'package:groovybox/ui/widgets/track_tile.dart';
import 'package:groovybox/ui/widgets/universal_image.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:material_symbols_icons/symbols.dart';
import 'package:styled_widget/styled_widget.dart';

class AlbumDetailScreen extends HookConsumerWidget {
  final AlbumData album;

  const AlbumDetailScreen({super.key, required this.album});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final repo = ref.watch(playlistRepositoryProvider.notifier);
    final tracksAsync = repo.watchAlbumTracks(album.album);

    // Responsive breakpoints
    final screenWidth = MediaQuery.sizeOf(context).width;
    final isLargeScreen = screenWidth > 900;

    return Scaffold(
      body: CustomScrollView(
        slivers: [
          // Full width app bar
          SliverAppBar(pinned: true, title: Text(album.album)),
          // Album cover section (full width on large screens, constrained on mobile)
          SliverToBoxAdapter(
            child: Container(
              decoration: BoxDecoration(
                borderRadius: const BorderRadius.all(Radius.circular(16)),
                boxShadow: [BoxShadow(blurRadius: 4)],
              ),
              constraints: const BoxConstraints(maxWidth: 320),
              child: AspectRatio(
                aspectRatio: 1,
                child: album.artUri != null
                    ? UniversalImage(uri: album.artUri!, fit: BoxFit.cover)
                    : Container(
                        color: Colors.grey[800],
                        child: const Icon(
                          Symbols.album,
                          size: 120,
                          color: Colors.white54,
                        ),
                      ),
              ).clipRRect(all: 16),
            ).center().padding(top: 16),
          ),
          // Content section with constrained width
          SliverToBoxAdapter(
            child: Align(
              alignment: Alignment.topCenter,
              child: ConstrainedBox(
                constraints: BoxConstraints(
                  maxWidth: isLargeScreen ? 800 : double.infinity,
                ),
                child: StreamBuilder<List<Track>>(
                  stream: tracksAsync,
                  builder: (context, snapshot) {
                    if (!snapshot.hasData) {
                      return const SizedBox(
                        height: 200,
                        child: Center(child: CircularProgressIndicator()),
                      );
                    }

                    final tracks = snapshot.data!;
                    if (tracks.isEmpty) {
                      return const SizedBox(
                        height: 200,
                        child: Center(child: Text('No tracks in this album')),
                      );
                    }

                    return Column(
                      children: [
                        // Action buttons
                        ConstrainedBox(
                          constraints: const BoxConstraints(maxWidth: 360),
                          child: Padding(
                            padding: EdgeInsets.all(16),
                            child: Column(
                              spacing: 12,
                              children: [
                                SizedBox(
                                  width: double.infinity,
                                  child: FilledButton.icon(
                                    onPressed: () {
                                      _playAlbum(ref, tracks);
                                    },
                                    icon: const Icon(Symbols.play_arrow),
                                    label: const Text('Play All'),
                                  ),
                                ),
                                SizedBox(
                                  width: double.infinity,
                                  child: OutlinedButton.icon(
                                    onPressed: () {
                                      _addToQueue(ref, tracks);
                                    },
                                    icon: const Icon(Symbols.queue_music),
                                    label: const Text('Add to Queue'),
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ),
                        // Track list
                        ...List.generate(tracks.length, (index) {
                          return _buildTrackTile(
                            ref,
                            tracks,
                            index,
                            isLargeScreen,
                          );
                        }),
                        // Gap for mini player
                        const Gap(80),
                      ],
                    );
                  },
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTrackTile(
    WidgetRef ref,
    List<Track> tracks,
    int index,
    bool isLargeScreen,
  ) {
    final track = tracks[index];
    return TrackTile(
      track: track,
      leading: Text(
        '${index + 1}'.padLeft(2, '0'),
        style: const TextStyle(color: Colors.grey, fontSize: 16),
      ).padding(right: isLargeScreen ? 24 : 16),
      showTrailingIcon: false,
      onTap: () {
        _playAlbum(ref, tracks, initialIndex: index);
      },
      padding: EdgeInsets.symmetric(
        horizontal: isLargeScreen ? 24 : 16,
        vertical: 8,
      ),
    );
  }

  void _playAlbum(WidgetRef ref, List<Track> tracks, {int initialIndex = 0}) {
    final loadingNotifier = ref.read(remoteTrackLoadingProvider.notifier);
    final audioHandler = ref.read(audioHandlerProvider);
    loadingNotifier.setLoading(true);
    audioHandler.playTracks(tracks, initialIndex: initialIndex).then((_) {
      loadingNotifier.setLoading(false);
    });
  }

  void _addToQueue(WidgetRef ref, List<Track> tracks) async {
    final audioHandler = ref.read(audioHandlerProvider);

    // Add tracks one by one to avoid interrupting playback
    for (final track in tracks) {
      try {
        final mediaItem = await _trackToMediaItem(track);
        await audioHandler.addQueueItem(mediaItem);
      } catch (e) {
        debugPrint('Error adding track ${track.title} to queue: $e');
      }
    }
  }

  Future<MediaItem> _trackToMediaItem(Track track) async {
    Uri? artUri;

    if (track.artUri != null) {
      // Check if it's a network URL or local file path
      if (track.artUri!.startsWith('http://') ||
          track.artUri!.startsWith('https://')) {
        // For remote tracks, skip artwork to avoid authentication issues
        // The artwork will be loaded separately when the track is played
        if (!track.path.startsWith('groovybox://')) {
          // Only try to cache artwork for non-remote tracks
          try {
            final cachedFile = await DefaultCacheManager().getSingleFile(
              track.artUri!,
            );
            artUri = Uri.file(cachedFile.path);
          } catch (e) {
            // If caching fails, skip artwork
            debugPrint('Failed to cache artwork for ${track.title}: $e');
          }
        }
        // For remote tracks, don't set artUri to avoid authentication issues
      } else {
        // It's a local file path
        artUri = Uri.file(track.artUri!);
      }
    }

    return MediaItem(
      id: track.path,
      album: track.album,
      title: track.title,
      artist: track.artist,
      duration: track.duration != null
          ? Duration(milliseconds: track.duration!)
          : null,
      artUri: artUri,
    );
  }
}

File: ui/screens/album_detail_screen.py
"""Album Detail Screen for GroovyBox.

This module displays the detail view for a specific album, showing
the album art, play controls, and a list of tracks in the album.
"""

import flet as ft
from data import playlist_repository as prepo
from logic.localize import tr
from ui.widgets.track_tile import TrackTile
from ui.widgets.universal_image import UniversalImage


class AlbumDetailView(ft.Container):
    """Album detail view showing album art and track listing.
    
    Provides controls to play all tracks or add them to the queue.
    Displays tracks with numbered leading indicators.
    
    Attributes:
        album: The AlbumData object being displayed.
    """

    def __init__(self, page: ft.Page, album):
        super().__init__()
        self._page = page
        self.album = album
        self._build()

    def _get_app(self):
        """Get the application instance from the session store."""
        return self._page.session.store.get("app")

    def _build(self):
        """Build the album detail view layout.
        
        Creates a scrollable view with:
        - Back button
        - Album art (240x240)
        - Play All and Add to Queue buttons
        - Numbered track list
        """
        tracks = prepo.watch_album_tracks(self.album.album)

        self.content = ft.Column(
            expand=True,
            scroll=ft.ScrollMode.AUTO,
            controls=[
                # Back navigation
                ft.Container(
                    padding=8,
                    content=ft.IconButton(ft.Icons.ARROW_BACK, on_click=lambda e: self._go_back()),
                ),
                # Album art
                ft.Container(
                    padding=16,
                    alignment=ft.Alignment(0, 0),
                    content=ft.Container(
                        width=240, height=240,
                        border_radius=16,
                        clip_behavior=ft.ClipBehavior.ANTI_ALIAS,
                        content=ft.Image(
                            src=self.album.art_uri,
                            fit=ft.BoxFit.COVER,
                            error_content=ft.Icon(ft.Icons.ALBUM, size=80, color=ft.Colors.WHITE54),
                        ),
                        bgcolor=ft.Colors.with_opacity(0.1, ft.Colors.ON_SURFACE),
                        shadow=ft.BoxShadow(blur_radius=6, color=ft.Colors.with_opacity(0.2, ft.Colors.SHADOW)),
                    ),
                ),
                # Action buttons
                ft.Container(
                    padding=ft.Padding(16, 0, 16, 0),
                    content=ft.Row(
                        tight=True,
                        alignment=ft.MainAxisAlignment.CENTER,
                        controls=[
                            ft.FilledButton(tr("playAll"), icon=ft.Icons.PLAY_ARROW, on_click=lambda e: self._play_all()),
                            ft.Container(width=12),
                            ft.OutlinedButton(tr("addToQueue"), icon=ft.Icons.QUEUE_MUSIC, on_click=lambda e: self._add_to_queue(tracks)),
                        ],
                    ),
                ),
                # Track list
                ft.Container(
                    expand=True,
                    padding=ft.Padding(16, 0, 16, 0),
                    content=ft.ListView(
                        spacing=4,
                        controls=[
                            TrackTile(
                                track=t,
                                leading=ft.Text(str(i + 1).zfill(2), color=ft.Colors.GREY, size=14),
                                on_tap=lambda e, idx=i: self._play_at(tracks, idx),
                                padding=4,
                            )
                            for i, t in enumerate(tracks)
                        ] + [ft.Container(height=80)],
                    ),
                ),
            ],
        )

    def _go_back(self):
        """Navigate back to the previous screen."""
        self._page.views.pop()
        self._page.run_task(self._page.push_route, self._page.views[-1].route if self._page.views else "/library")

    def _play_all(self, tracks=None, initial_index=0):
        """Play all tracks in the album.
        
        Args:
            tracks: Optional pre-loaded track list. If None, fetches from DB.
            initial_index: Index to start playback from.
        """
        if tracks is None:
            tracks = prepo.watch_album_tracks(self.album.album)
        if tracks:
            app = self._get_app()
            if app:
                app.audio_player.play_tracks(tracks, initial_index)

    def _play_at(self, tracks, idx):
        """Play the album starting from a specific track index.
        
        Args:
            tracks: List of tracks in the album.
            idx: Index of the track to start from.
        """
        self._play_all(tracks, idx)

    def _add_to_queue(self, tracks):
        """Add all album tracks to the playback queue.
        
        Args:
            tracks: List of tracks to add.
        """
        app = self._get_app()
        if app and tracks:
            for t in tracks:
                app.audio_player.queue.append(t)
            if app.audio_player.on_queue_change:
                app.audio_player.on_queue_change()
